libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -0,0 +1,439 @@
+
+	bool const ignoreMemory = MSizeFinder::containsMSize(_context.dialect, _ast);
+	UnusedStoreEliminator rse{
+		_context.dialect,
+		functionSideEffects,
+		ControlFlowSideEffectsCollector{_context.dialect, _ast}.functionSideEffectsNamed(),
+		values,
+		ignoreMemory
+	};
+	rse(_ast);
+	rse.changeUndecidedTo(State::Unused, Location::Memory);
+	rse.changeUndecidedTo(State::Used, Location::Storage);
+	rse.scheduleUnusedForDeletion();
+
+	StatementRemover remover(rse.m_pendingRemovals);
+	remover(_ast);
+}
+
+void UnusedStoreEliminator::operator()(FunctionCall const& _functionCall)
+{
+	UnusedStoreBase::operator()(_functionCall);
+
+	for (Operation const& op: operationsFromFunctionCall(_functionCall))
+		applyOperation(op);
+
+	ControlFlowSideEffects sideEffects;
+	if (auto builtin = m_dialect.builtin(_functionCall.functionName.name))
+		sideEffects = builtin->controlFlowSideEffects;
+	else
+		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
+
+	if (!sideEffects.canContinue)
+	{
+		changeUndecidedTo(State::Unused, Location::Memory);
+		changeUndecidedTo(sideEffects.canTerminate ? State::Used : State::Unused, Location::Storage);
+	}
+}
+
+void UnusedStoreEliminator::operator()(FunctionDefinition const& _functionDefinition)
+{
+	ScopedSaveAndRestore storeOperations(m_storeOperations, {});
+	UnusedStoreBase::operator()(_functionDefinition);
+}
+
+
+void UnusedStoreEliminator::operator()(Leave const&)
+{
+	changeUndecidedTo(State::Used);
+}
+
+void UnusedStoreEliminator::visit(Statement const& _statement)
+{
+	using evmasm::Instruction;
+
+	UnusedStoreBase::visit(_statement);
+
+	auto const* exprStatement = get_if<ExpressionStatement>(&_statement);
+	if (!exprStatement)
+		return;
+
+	FunctionCall const* funCall = get_if<FunctionCall>(&exprStatement->expression);
+	if (!funCall)
+		return;
+	optional<Instruction> instruction = toEVMInstruction(m_dialect, funCall->functionName.name);
+	if (!instruction)
+		return;
+
+	if (!ranges::all_of(funCall->arguments, [](Expression const& _expr) -> bool {
+		return get_if<Identifier>(&_expr) || get_if<Literal>(&_expr);
+	}))
+		return;
+
+	if (
+		*instruction == Instruction::SSTORE ||
+		(!m_ignoreMemory && (
+			*instruction == Instruction::EXTCODECOPY ||
+			*instruction == Instruction::CODECOPY ||
+			*instruction == Instruction::CALLDATACOPY ||
+			*instruction == Instruction::RETURNDATACOPY ||
+			*instruction == Instruction::MSTORE ||
+			*instruction == Instruction::MSTORE8
+		))
+	)
+	{
+		m_stores[YulString{}].insert({&_statement, State::Undecided});
+		vector<Operation> operations = operationsFromFunctionCall(*funCall);
+		yulAssert(operations.size() == 1, "");
+		m_storeOperations[&_statement] = move(operations.front());
+	}
+}
+
+void UnusedStoreEliminator::finalizeFunctionDefinition(FunctionDefinition const&)
+{
+	changeUndecidedTo(State::Used);
+	scheduleUnusedForDeletion();
+}
+
+vector<UnusedStoreEliminator::Operation> UnusedStoreEliminator::operationsFromFunctionCall(
```

Comment:
I wonder whether we should move the main part of this to some abstract semantic property as well.
I.e. just have the builtins tell us "I'm reading the memory range [arg0, arg1)" instead of just "I'm reading memory".

Reply:
It might be useful to have that information elsewhere and also if e.g. there's ever a new EVM opcode that accesses memory or storage, we'll definitely define its semantic properties, but it's much easier to miss editing stuff elsewhere like here.

Reply:
Yep, true! We could use that routine in/as replacement for `isSimpleLoad` / `isSimpleStore`